lib/srv/forward/sshserver.go

```diff
@@ -851,6 +850,84 @@ func (s *Server) handleConnection(ctx context.Context, chans <-chan ssh.NewChann
 	}
 }
 
+// handleClientChannels handles channel open requests from the remote server.
+func (s *Server) handleClientChannels(ctx context.Context) {
+	forwardedTCPIP := s.remoteClient.HandleChannelOpen(teleport.ChanForwardedTCPIP)
+
+	for {
+		select {
+		case nch := <-forwardedTCPIP:
+			chanCtx, nch := tracessh.ContextFromNewChannel(nch)
+			ctx, span := s.tracerProvider.Tracer("ssh").Start(
+				oteltrace.ContextWithRemoteSpanContext(ctx, oteltrace.SpanContextFromContext(chanCtx)),
+				fmt.Sprintf("ssh.Forward.OpenChannel/%s", nch.ChannelType()),
+				oteltrace.WithSpanKind(oteltrace.SpanKindServer),
+				oteltrace.WithAttributes(
+					semconv.RPCServiceKey.String("ssh.ForwardServer"),
+					semconv.RPCMethodKey.String("OpenChannel"),
+					semconv.RPCSystemKey.String("ssh"),
+				),
+			)
+
+			go func() {
+				defer span.End()
+				s.handleForwardedTCPIPRequest(ctx, nch)
+			}()
+		case <-ctx.Done():
```

Comment:
The `NewChannel` and `Request` channels "must be serviced or the connection will hang", as per the `golang.org/x/crypto/ssh` docs. We should not exit and leave the channel hanging just because an asynchronous cancellation request has come through.

Reply:
The channel won't be left hanging, it'll be closed by `utils.ProxyConn` when the context is canceled.

Reply:
I'm talking about the go channels; if the ssh channel will be closed when the context is canceled then it's even more reason to not early exit on context cancel here. This whole loop should just be a single range over `forwardedTCPIP`:
```go
	for nch := range forwardedTCPIP {
		ctx, span := ...
		go func() {
			defer span.End()
			s.handleForwardedTCPIPRequest(ctx, nch)
		}()
	}
```